windows test failed:
```
[2026-05-12T10:57:36.000Z] [isolated][googletest] ceph_test_libcephfs_snapdiff failed. Error: Traceback (most recent call last):
[2026-05-12T10:57:36.000Z] [isolated][googletest] ceph_test_libcephfs_suidsgid passed.
[2026-05-12T10:57:36.000Z] [isolated][googletest] ceph_test_libcephfs_vxattr passed.
[2026-05-12T10:57:36.000Z] [isolated][googletest] unittest_admin_socket passed.
unittest_global_doublefree
ceph_test_libcephfs_snapdiff
The process tried to write to a nonexistent pipe.
[2026-05-12T10:57:40.000Z] subunit-trace reports failures: Command failed: type C:\workspace\test_results\subunit.out | subunit-trace
One or more test suites have failed
At C:\workspace\repos\ceph-win32-tests\test_host\run_tests.ps1:28 char:9
+         throw "One or more test suites have failed"
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (One or more test suites have failed:String) [], RuntimeException
    + FullyQualifiedErrorId : One or more test suites have failed
 
+ EXIT_CODE=1
+ [[ 1 -eq 124 ]]
+ return 1
```

--- a/src/crimson/osd/osd_operations/pg_advance_map.cc
+++ b/src/crimson/osd/osd_operations/pg_advance_map.cc
@@ -59,7 +59,6 @@ PGPeeringPipeline &PGAdvanceMap::peering_pp(PG &pg)
 seastar::future<> PGAdvanceMap::start()
 {
   LOG_PREFIX(PGAdvanceMap::start);
-  using cached_map_t = OSDMapService::cached_map_t;
 
   DEBUG("{}: start", *this);
   auto exit_handle = seastar::defer([FNAME, thiz = IRef{this}] {
@@ -112,7 +111,6 @@ seastar::future<> PGAdvanceMap::check_for_splits(
     cached_map_t next_map)
 {
   LOG_PREFIX(PGAdvanceMap::check_for_splits);
-  using cached_map_t = OSDMapService::cached_map_t;
   cached_map_t old_map = co_await shard_services.get_map(old_epoch);
   if (!old_map->have_pg_pool(pg->get_pgid().pool())) {
     DEBUG("{} pool doesn't exist in epoch {}", pg->get_pgid(),

--- a/src/crimson/osd/osd_operations/pg_advance_map.cc
+++ b/src/crimson/osd/osd_operations/pg_advance_map.cc
@@ -35,7 +35,6 @@ void PGAdvanceMap::print(std::ostream &lhs) const
 {
   lhs << "PGAdvanceMap("
       << "pg=" << pg->get_pgid()
-      << " from=" << (from ? *from : -1)
       << " to=" << to;
   if (do_init) {
     lhs << " do_init";
@@ -47,9 +46,6 @@ void PGAdvanceMap::dump_detail(Formatter *f) const
 {
   f->open_object_section("PGAdvanceMap");
   f->dump_stream("pgid") << pg->get_pgid();
-  if (from) {
-    f->dump_int("from", *from);
-  }
   f->dump_int("to", to);
   f->dump_bool("do_init", do_init);
   f->close_section();
@@ -77,23 +73,13 @@ seastar::future<> PGAdvanceMap::start()
       return seastar::now();
     }
 
-    /*
-     * PGAdvanceMap is scheduled at pg creation and when
-     * broadcasting new osdmaps to pgs. We are not able to serialize
-     * between the two different PGAdvanceMap callers since a new pg
-     * will get advanced to the latest osdmap at it's creation.
-     * As a result, we may need to adjust the PGAdvance operation
-     * 'from' epoch.
-     * See: https://tracker.ceph.com/issues/61744
-     */
-    from = pg->get_osdmap_epoch();
+    epoch_t from = pg->get_osdmap_epoch();
     if (do_init) {
       pg->handle_initialize(rctx);
       pg->handle_activate_map(rctx);
     }
-    ceph_assert(std::cmp_less_equal(*from, to));
     return seastar::do_for_each(
-      boost::make_counting_iterator(*from + 1),
+      boost::make_counting_iterator(from + 1),
       boost::make_counting_iterator(to + 1),
       [this, FNAME](epoch_t next_epoch) {
 	DEBUG("{}: start: getting map {}",

--- a/src/crimson/osd/osd_operations/pg_advance_map.cc
+++ b/src/crimson/osd/osd_operations/pg_advance_map.cc
@@ -62,55 +62,49 @@ seastar::future<> PGAdvanceMap::start()
   using cached_map_t = OSDMapService::cached_map_t;
 
   DEBUG("{}: start", *this);
-
-  IRef ref = this;
-  return enter_stage<>(
-    peering_pp(*pg).process
-  ).then([this, FNAME] {
-    // pg may have been deleted while this op was queued; see PG::do_delete_work.
-    if (pg->is_deleted()) {
-      DEBUG("{}: pg is deleted, skipping advance", *this);
-      return seastar::now();
-    }
-
-    epoch_t from = pg->get_osdmap_epoch();
-    if (do_init) {
-      pg->handle_initialize(rctx);
-      pg->handle_activate_map(rctx);
-    }
-    return seastar::do_for_each(
-      boost::make_counting_iterator(from + 1),
-      boost::make_counting_iterator(to + 1),
-      [this, FNAME](epoch_t next_epoch) {
-	DEBUG("{}: start: getting map {}",
-		       *this, next_epoch);
-	return shard_services.get_map(next_epoch).then(
-	  [this, FNAME] (cached_map_t&& next_map) {
-	    DEBUG("{}: advancing map to {}",
-		  *this, next_map->get_epoch());
-        // Use the current OSDMap epoch to check for splits consecutively.
-        epoch_t current_epoch = pg->get_osdmap_epoch();
-        pg->handle_advance_map(next_map, rctx);
-        DEBUG("{}: checking for splits between {} and {}",
-              *this, current_epoch, next_map->get_epoch());
-        return check_for_splits(current_epoch, next_map);
-	  });
-      }).then([this, FNAME] {
-	pg->handle_activate_map(rctx);
-	DEBUG("{}: map activated", *this);
-	if (do_init) {
-	  shard_services.pg_created(pg->get_pgid(), pg);
-	  INFO("PGAdvanceMap::start new pg {}", *pg);
-	}
-	return pg->complete_rctx(std::move(rctx));
-      });
-  }).then([this, FNAME] {
-    DEBUG("{}: complete", *this);
-    return handle.complete();
-  }).finally([this, FNAME, ref=std::move(ref)] {
-    DEBUG("{}: exit", *this);
-    handle.exit();
+  auto exit_handle = seastar::defer([FNAME, thiz = IRef{this}] {
+    DEBUG("{}: exit", *thiz);
+    thiz->handle.exit();
   });
+
+  co_await enter_stage<>(peering_pp(*pg).process);
+
+  // pg may have been deleted while this op was queued; see PG::do_delete_work.
+  if (pg->is_deleted()) {
+    DEBUG("{}: pg is deleted, skipping advance", *this);
+    co_return;
+  }
+
+  epoch_t from = pg->get_osdmap_epoch();
+  if (do_init) {
+    pg->handle_initialize(rctx);
+    pg->handle_activate_map(rctx);
+  }
+  for (epoch_t next_epoch = from + 1; next_epoch <= to; next_epoch++) {
+    DEBUG("{}: start: getting map {}",
+          *this, next_epoch);
+    cached_map_t next_map = co_await shard_services.get_map(next_epoch);
+    DEBUG("{}: advancing map to {}",
+          *this, next_map->get_epoch());
+    // Use the current OSDMap epoch to check for splits consecutively.
+    epoch_t current_epoch = pg->get_osdmap_epoch();
+    pg->handle_advance_map(next_map, rctx);
+    DEBUG("{}: checking for splits between {} and {}",
+          *this, current_epoch, next_map->get_epoch());
+    co_await check_for_splits(current_epoch, std::move(next_map));
+  }
+
+  pg->handle_activate_map(rctx);
+  DEBUG("{}: map activated", *this);
+  if (do_init) {
+    shard_services.pg_created(pg->get_pgid(), pg);
+    INFO("PGAdvanceMap::start new pg {}", *pg);
+  }
+  co_await pg->complete_rctx(std::move(rctx));
+  DEBUG("{}: complete", *this);
+
+  co_await handle.complete();
+  exit_handle.cancel();
 }
 
 seastar::future<> PGAdvanceMap::check_for_splits(

--- a/src/crimson/osd/osd_operations/pg_advance_map.h
+++ b/src/crimson/osd/osd_operations/pg_advance_map.h
@@ -31,7 +31,6 @@ protected:
   ShardServices &shard_services;
   PipelineHandle handle;
 
-  std::optional<epoch_t> from;
   epoch_t to;
 
   PeeringCtx rctx;

